(Design) Number Selection (#471)
unicode-org/message-format-wg: exploration/number-selection.md

## Constraints

ICU MF1 messages using `plural` and `selectordinal` should be representable in MF2.

## Proposed Design

Given that we already have a `:number`,
it makes sense to add a `<matchSignature>` to it with an option

```xml
<option name="select" values="plural ordinal exact" default="plural" />
```

The default `plural` value is presumed to be the most common use case,
and it affords the least bad fallback when used incorrectly:
Using "plural" for "exact" still selects exactly matching cases,
whereas using "exact" for "plural" will not select LDML category matches.
This might not be noticeable in the source language,
but may cause problems in target languages that the original developer is not considering.

Additional options such as `minimumFractionDigits` and others already supported by `:number`
should also be supported.

If PR [#532](https://github.com/unicode-org/message-format-wg/pull/532) is accepted,
also add the following `<alias>` definitions to `<function name="number">`:

```xml
<alias name="plural" supports="match">
	<setOption name="select" value="plural"/>
</alias>
<alias name="ordinal" supports="match">
	<setOption name="select" value="ordinal"/>
</alias>

> ```

Reviewer: Is this list exhaustive? #532 includes `:integer`. Presumably `:integer` is an integer plural selector by default (in addition to being a formatter).

Other candidates for shorthands might be `:percent`, `:currency` and (maybe??) `:scientific`. The first two are shorthands in ICU4J MF1. `scientific` is a numeric formatting option (but not exposed by MF1). 

There is also the potentially lamentable `:spellout` from ICU4J MF1.

I'm guessing that what we want to promote is single-annotation messages, e.g.:

```
.input {$var :spellout}
.match {$var}                               // it's already a plural
[0]   {{You have no more bone dragons}}
[one] {{You have {$var} more bone dragon}}  // "You have one more..."
[*]   {{You have {$var} more bone dragons}}
```

Author: I think the ones you mention are all formatters? The list here is purely about `.match` selectors, and for that I think just `:plural` and `:ordinal` could be enough. If you need exact matching, then the relatively rare `:number select=exact` is pretty good?

Author: To clarify a bit, a proposed alias like `:integer` or `:spellout` still ends up calling the `:number` function, and so if the expression doesn't have an explicit `select` attribute and the alias does not set `supports="format"`, it'll end up inheriting the default `select="plural"` for selection, together with the rest of the expression & alias options.

Author: Here's how I would classify the different aliases that have been mentioned so far:
- `:integer` -- The most obvious one, and therefore included already in [#532](https://github.com/unicode-org/message-format-wg/pull/532).
- `:plural` and `:ordinal` -- Functions for selection only with `supports="match"`. Included in this design doc.
- `:percent`, `:currency`,`:scientific`, `:spellout` -- Possible aliases for later/separate consideration. As with `:integer`, they might be used also for selection, but I don't see how any of them would specify a `<setOption name="select">`. So I don't think they belong in a design doc on Number Selection.

Once #532 lands, would a PR adding that third set be the right focal point for discussion of their merits, rather than this PR?

Author: I agree that `:ordinal` could also be a formatter, but I am not here proposing that it be one of the core formatters that we expect all MF2 implementations to provide. That could be considered as a separate further step, should this design on number _selection_ first introduce it as a selector.

Author: I would very strongly prefer taking this smaller step first, and not expanding the scope of this design document.

# Selection on Numerical Values

Status: **Accepted**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@eemeli</dd>
		<dt>First proposed</dt>
		<dd>2023-09-06</dd>
		<dt>Pull Request</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/471">#471</a></dd>
	</dl>
</details>

## Objective

Define how selection on numbers happens.

## Background

> As discussed on numerous past occasions and explicitly identified in
> <a href="https://github.com/unicode-org/message-format-wg/pull/457">#457</a>,
> it would be good for both plural and non-plural match selection to be supported in MF2.

Reviewer: This is slightly misleading--it implies a binary plural/non-plural set of selection, whereas we go on to point out that there are a variety of number selection needs. Perhaps:

```suggestion
As discussed by the working group and explicitly identified in
<a href="https://github.com/unicode-org/message-format-wg/pull/457">discussion of #457</a>,
there is a need to support multiple types of selection on numeric values in MF2.
```

MF1 supported selection on either cardinal plurals or ordinal numbers,
via the `plural` and `selectordinal` selectors.
It also customized this selection beyond the capabilities of `com.ibm.icu.text.PluralRules`
by allowing for explicit value matching and an `offset` parameter.

As pointed out by <a href="https://github.com/mihnita">@mihnita</a> in particular,
category selection is not always appropriate for selection on a number:
the number may be representing some completely other quantity,
such as a four-digit year or the integer value of an enumerator.

Furthermore, as pointed out by <a href="https://github.com/ryzokuken">@ryzokuken</a>
in <a href="https://github.com/unicode-org/message-format-wg/pull/457#discussion_r1307443288">#457 (comment)</a>,

> ordinal selection is similar to plural selection, but not identical.

Reviewer: ```suggestion
ordinal selection works similarly to plural selection, but uses a separate set of rules
for each locale.
This is visible in English, for example, where plural rules use only `one` and `other`
but ordinal rules use `one` (_1st_, _21st_, etc.), `few` (_2nd_, _22nd_, etc.), 
`many` (_3rd_, _23rd_, etc.), and `other` (_4th_, _5th_, etc.).
```

> Additionally,
> MF1 provides `ChoiceFormat` selection based on whether a number is within a range,
> and both JS and ICU PluralRules implementations provide for determining the plural category
> of a range based on its start and end values.
> These range-based selectors are not initially considered here.

Reviewer: These are separate things; they should be kept separate.

```suggestion
Additionally,
MF1 provides `ChoiceFormat` selection based on a complex rule set
(and which allows determining if a number falls into a specific range).

Both JS and ICU PluralRules implementations provide for determining the plural category
of a range based on its start and end values.
Range-based selectors are not initially considered here.
```

but can cause problems in target locales that the original developer is not considering.

> For example, a naive developer might use a special message for the value `1` without
> considering other locale's need for a `one` plural:
>
> ```
> .match {$var}
> [1] {{You have one last chance}}
> [one] {{You have {$var} chance remaining}} // needed by languages such as Polish or Russian
> [*] {{You have {$var} chances remaining}}
> ```

Additional options such as `minimumFractionDigits` and others already supported by `:number`
should also be supported.

If PR [#532](https://github.com/unicode-org/message-format-wg/pull/532) is accepted,
also add the following `<alias>` definitions to `<function name="number">`:

```xml
<alias name="plural" supports="match">
	<setOption name="select" value="plural"/>
</alias>
<alias name="ordinal" supports="match">
	<setOption name="select" value="ordinal"/>
</alias>
```

## Alternatives Considered

### Completely Separate Functions

An alternative approach to this problem could be to leave the `:number` `<matchSignature>` undefined,
and to define three further functions, each with a `<matchSignature>`:

- `:plural`
- `:ordinal`

> - `:exact`

Author: ```suggestion
- `:exact` (actual name TBD, pending the resolution of [#433](https://github.com/unicode-org/message-format-wg/issues/433)
```

which would each need the same set of options as `:number`, except for `type`.

This approach would also mostly work, but it introduces new failure modes:

- If a `:number` is used directly as a selector, this should produce a runtime error.
- If a `:plural`/`:ordinal`/`:exact` is used as a formatter, this should produce a runtime error.
- Developers are less encouraged to use the same formatting and selection options.

To expand on the last of these,
consider this message:

```
.match {$count :plural minimumFractionDigits=1}
0 {{You have no apples}}
1 {{You have exactly one apple}}
* {{You have {$count :number minimumFractionDigits=1} apples}}
```

Here, because selection on `:number` is not allowed,
it's easy to duplicate the options because _some_ annotation is required on the selector.
It would also be relatively easy to leave out the `minimumFractionDigits=1` option from the selector,
as it's not required for the English source.

With the proposed design, this message would much more naturally be written as:

```
.input {$count :number minimumFractionDigits=1}
.match {$count}
0 {{You have no apples}}
1 {{You have exactly one apple}}

> * {{You have {$count} apples}}

Reviewer: As noted before we should model good behavior. This message should also have a `one` rule

```suggestion
one {{You have {$count} apple}}
* {{You have {$count} apples}}
```

a numeric value.

As a user, I want to write messages that mix exact matching and 
either plural or ordinal selection in a single message. 
> For example:
>```
>.match {$numRemaining}
>0 {{You have no more chances remaining (exact match)}}
>1 {{You have one more chance remaining (exact match)}}
>one {{You have {$numRemaining} chance remaining (plural)}}
> * {{You have {$numRemaining} chances remaining (plural)}}
>```

## Requirements

- Enable cardinal plural selection.
- Enable ordinal number selection.
- Enable exact match selection.
- Enable simultaneous "exact match" and either plural or ordinal selection.
  > For example, allow variants `[1]` and `[one]` in the same message.
- Selection must support relevant formatting options, such as `minimumFractionDigits`.
  > For example, in English the value `1` matches plural rule `one` while the value `1.0`
  > matches the plural rule `other` (`*`)
- Encourage developers to provide the formatting options used in patterns to the selector
  so that proper selection can be done.

## Constraints

ICU MF1 messages using `plural` and `selectordinal` should be representable in MF2.

## Proposed Design

> Given that we already have a `:number`,

Reviewer 2: > Given that we already have a `:number`,
> it makes sense to add a `<matchSignature>` to it with an option

Sorry, but it is not that obvious.
`:number` is not a question, so one can't decide on it.

You have to ask a question about the number.

"Hey, `:number 43` , true or false?
If you answer correctly you win 1000$"
That means nothing. It is not a good question.

`is :number 43 even`" is a question, you can answer.
`is :number 43 prime`" is a question, you can answer.
`is :number 43 plural`" is a question, you can answer.

Reviewer: There are two questions here.

The first is: should the `:number` function used for formatting also do selection? @eemeli has already pointed out that we want users to use the same configuration for selection (e.g. with `:plural`) as will ultimately be used to format the number. This message produces grammatically inferior results:

```
.local $tenths = {$num :number minFractionDigits=1}
.match {$num :plural}
one {{You have {$tenths} value}}  // won't be seen
* {{You have {$tenths} values}}
```

The second question is: if `:number` is a selector, what sort of selector is it by default? @eemeli answers that `:plural` is the most common usage. Alternatives would be `:ordinal` and whatever we call exact match. I agree with @eemeli that `:plural` is the far-and-away most common case and thus should be the default.

So... @mihnita are you objecting to the phrasing here or the conclusion? If the phrasing, do you want to suggest alternate text?

Author: @aphillips Your "won't be seen" comment is a bit misleading, as the problem with that message is that the `one` variant _may_ indeed be seen, formatted as "You have 1.0 value".

Reviewer: You're right.. that comment goes against what I'm trying to say, so I fixed it.

Reviewer 2: Since we already defined `$tenths` and use it in the body of the options, then I expect `:plural` on it:
```
.local $tenths = {$num :number minFractionDigits=1}
.match {$tenths :plural}
one {{You have {$tenths} value}}  // not desirable
* {{You have {$tenths} values}}
```

And not doing that is easy to do at lint time.
